### ISSUE TITLE
Rule for reference checking in KDoc comments

### ISSUE THREAD

**Issue by utybo** [detekt/detekt#3657]
## Expected Behavior of the rule

Report invalid references in KDoc comments with the `[]` syntax.

### OK situations

Consider that this is the only file in our project for the `some.library` package.

```kotlin
package some.library

class SomeClass { /* ... */ }

interface SomeInterface { /* ... */ }

/**
 * Does something, use in the context of [SomeClass], using [some.library.SomeInterface]
 */
fun someFunction() {
    TODO()
}
```

### Invalid References

IntelliJ properly reports both of the references with a warning, and detekt should mirror that behavior:

```kotlin
package some.library

class SomeClass { /* ... */ }

interface SomeInterface { /* ... */ }

/**
 * Does something, use in the context of [SomeAlien], using [some.library.DoesNotExist]
 *                                       ^^^^^^^^^^^        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
 */
fun someFunction() {
    TODO()
}
```

## Context

This rule should report invalid references within KDoc comments. These comments can reference other elements (such as functions, classes or properties) [using the `[x]` syntax](https://kotlinlang.org/docs/kotlin-doc.html#links-to-elements), where `x` is the name of the referenced element. IntelliJ properly warns of invalid references, such as this one...

![image](https://user-images.githubusercontent.com/5175705/113693843-54cc3d80-96cf-11eb-81b9-d62dad922c8f.png)

...but Detekt does not seem to have a rule for this. This would be invaluable to report KDoc comments that have not been updated after some refactoring session and that point to now invalid elements.

Note that this feature is quite complex, as it also properly handles imports...

> Names in links are resolved using the same rules as if the name was used inside the element being documented. In particular, this means that if you have imported a name into the current file, you don't need to fully qualify it when you use it in a KDoc comment.

... and overloads...

> Note that KDoc does not have any syntax for resolving overloaded members in links. Since the Kotlin documentation generation tool puts the documentation for all overloads of a function on the same page, identifying a specific overloaded function is not required for the link to work.

See [the official documentation on this](https://kotlinlang.org/docs/kotlin-doc.html#links-to-elements)
